qmt/geometry/geo_3d_data.py

```diff
@@ -0,0 +1,367 @@
+        return file_path
+
+    def xsec_to_2d(self, xsec_name: str, lunit: Optional[str] = None) -> Geo2DData:
+        """Generates a Geo2DData from a cross section
+
+        Parameters
+        ----------
+        xsec_name : str
+            Name of the cross section
+        lunit : Optional[str] :
+            (Default value = None)
+        lunit: Optional[str] :
+             (Default value = None)
+        Returns
+        -------
+        None
+
+        """
+
+        # Get our new coordinates
+        # This constructions tries to align the new coordinates to our old coordinates
+        # In particular, the map from projection axis -> new axes is
+        # [1,0,0] -> [0,1,0] [0,0,1]
+        # [0,1,0] -> [1,0,0] [0,0,1]
+        # [0,0,1] -> [1,0,0] [0,1,0]
+
+        # Find out which axis the projection axis is most closely aligned to
+        x_new = np.array(self.xsecs[xsec_name]["axis"])
+        ind = np.argmax(np.abs(x_new))
+        y_new = np.array([0, 1.0, 0]) if ind == 0 else np.array([1.0, 0, 0])
+        y_new -= y_new.dot(x_new) * x_new
+        z_new = np.cross(x_new, y_new)
+        # Adjust our second axis so that the "height axis" is correctly aligned
+        if z_new[2] < 0:
+            z_new = -z_new
+
+        def _project(vec):
+            """Projects a 3D vector into our 2D cross section plane
+            
+            Parameters
+            ----------
+            vec :
+            Returns
+            -------
+            List of projection.
+
+            """
+            vec = vec - x_new * self.xsecs[xsec_name]["distance"]
+            return [vec.dot(y_new), vec.dot(z_new)]
+
+        def _inverse_project(vec):
+            """Inverse of _project
+            
+            Parameters
+            ----------
+            vec :
+            Returns
+            -------
+            Float, inverse of _project.
+
+            """
+            return (
+                x_new * self.xsecs[xsec_name]["distance"]
+                + vec[0] * y_new
+                + vec[1] * z_new
+            )
+
+        part_polygons = {}
+        virtual_part_polygons = {}
+        # part_polygons is a dictionary of part_name to polygons. virtual_part_polygons
+        # is the same for virtual parts
+        for part_name in self.build_order:
+            polygons = []
+            for name, points in self.xsecs[xsec_name]["polygons"].items():
+                if name.startswith(f"{part_name}_"):
+                    poly = Polygon(map(_project, points))
+                    # we add a name to the polygon so we can reference it easier later
+                    poly.name = name
+                    polygons.append(poly)
+            if polygons:
+                if self.parts[part_name].virtual:
+                    virtual_part_polygons[part_name] = polygons
+                else:
+                    part_polygons[part_name] = polygons
+
+        def _build_containment_graph(poly_list):
```

Comment:
TODO: this should be tested.

Reply:
There're no unit tests for this function, but it is tested as part of `test_xsection` and `test_simple_xsection`. There're a lot of functions that have no unit tests to them. Do you think this function looks particularly suspicious?